pkg/terraformer/terraformer.go

```diff
@@ -205,3 +221,65 @@ func (t *Terraformer) executeTerraform(ctx context.Context, command Command) err
 	log.Info("terraform process finished successfully", "command", command)
 	return nil
 }
+
+func (t *Terraformer) addFinalizer(ctx context.Context) error {
+	logger := t.log.WithName("add-finalizer")
+	return t.updateObjectFinalizers(ctx, logger, controllerutil.AddFinalizer)
+
+}
+
+func (t *Terraformer) removeFinalizer(ctx context.Context) error {
+	logger := t.log.WithName("remove-finalizer")
+	return t.updateObjectFinalizers(ctx, logger, controllerutil.RemoveFinalizer)
+}
+
+func (t *Terraformer) updateObjectFinalizers(ctx context.Context, log logr.Logger, patchObj func(obj controllerutil.Object, finalizerName string)) error {
+	objects := []controllerutil.Object{
+		&corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: t.config.Namespace,
+				Name:      t.config.VariablesSecretName,
+			},
+		},
+		&corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: t.config.Namespace,
+				Name:      t.config.ConfigurationConfigMapName,
+			},
+		},
+		&corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: t.config.Namespace,
+				Name:      t.config.StateConfigMapName,
+			},
+		},
```

Comment:
If the object is missing, it will be skipped. Isn't this enough?
Usually, this is not the case in the Gardener scenario because the provider extension creates these resources before to execute the terraformer - the `CreateState` [here](https://github.com/gardener/gardener-extension-provider-aws/blob/d91cadaa35c9b73f1fbc78743772870fdd24b420/pkg/controller/infrastructure/actuator_reconcile.go#L45) takes care to create configmap with empty state.

Reply:
Hmm, the problem is, that if the state ConfigMap doesn't exist before terraformer is executed, it will skip adding its finalizer and will create it once there was an update to the state file. The result is, that there is some state saved in the state configmap, but there is no finalizer on it.

I would suggest the following:
If the state configmap doesn't exist when trying to add a finalizer to it, we create an empty state configmap with the finalizer on it, basically [here](https://github.com/gardener/terraformer/pull/65/files#diff-ae1251668ebb045109ad93128a7c02beab02a6b502f60704a9332cde7b6b7588R299).
WDYT?

Reply:
Sounds good to me, creating empty resource with no data should be fine.

Reply:
Done.